### PR TITLE
Add `createNamedParameter` to `QueryBuilder`

### LINKED ISSUE #9878: QueryBuilder function "createNamedParameter" like in doctrine/dbal
### Feature Request

Hello,
i noticed that the function "createNamedParameter" does not exist in the [QueryBuilder](https://github.com/doctrine/orm/blob/2.12.x/lib/Doctrine/ORM/QueryBuilder.php), unlike the [QueryBuilder](https://github.com/doctrine/dbal/blob/3.3.x/src/Query/QueryBuilder.php#L1463) from "doctrine/dbal".
Is there a reason why it is not available in "doctrine/orm"?
Otherwise I can create a pull request.

|    Q        |   A
|------------ | ------
| New Feature | yes
| RFC         | no
| BC Break    | no

#### Summary

Documentation of the function "createNamedParameter": https://www.doctrine-project.org/projects/doctrine-dbal/en/latest/reference/query-builder.html#binding-parameters-to-placeholders

### PR DESCRIPTION
**Pull request by norkunas**
Closes #9878

Replaces #9883

Cherry picked tests from #9883

Fixed review suggestions

### CODE REVIEW

**Comment by stof** [comment]
I don't think `createPositionalParameter` makes sense, as the index of the positional parameter depends on its position in the final query, which means that this API will very easily get broken. You cannot use positional parameters without caring about all of them.
If you don't want to care about the exact names of your parameters, there is no need to make them positional instead of making them named as `dcValue<increment>`

**Comment by norkunas** [comment]
> I don't think `createPositionalParameter` makes sense, as the index of the positional parameter depends on its position in the final query, which means that this API will very easily get broken. You cannot use positional parameters without caring about all of them.
> If you don't want to care about the exact names of your parameters, there is no need to make them positional instead of making them named as `dcValue<increment>`

I care about only named parameters, so this one could be removed in my opinion

**Comment by greg0ire** [comment]
Sorry for originally suggesting adding it, I didn't see the flaw. Please remove it, and squash both commits together.

File: docs/en/reference/query-builder.rst [hunk 1]
```diff
@@ -611,3 +611,28 @@ same query of example 6 written using
       ->add('from', new Expr\From('User', 'u'))
       ->add('where', new Expr\Comparison('u.id', '=', '?1'))
       ->add('orderBy', new Expr\OrderBy('u.name', 'ASC'));
+
+Binding Parameters to Placeholders
+----------------------------------
+
+It is often not necessary to know about the exact placeholder names
+during the building of a query. You can use two helper methods
```
Comment (greg0ire) [inline docs/en/reference/query-builder.rst]:
```suggestion
when building a query. You can use two helper methods
```

File: src/QueryBuilder.php [hunk 2]
```diff
@@ -1336,6 +1343,71 @@ public function resetDQLPart(string $part): static
...
+     * @param ParameterType|ArrayParameterType|string|int|null $type        ParameterType::*, ArrayParameterType::* or \Doctrine\DBAL\Types\Type::* constant
+     * @param non-empty-string|null                            $placeholder The name to bind with. The string must start with a colon ':'.
+     *
+     * @return non-empty-string the placeholder name used.
+     */
+    public function createNamedParameter(mixed $value, ParameterType|ArrayParameterType|string|int|null $type = null, string|null $placeholder = null): string
+    {
+        if ($placeholder === null) {
+            $this->boundCounter++;
+            $placeholder = ':dcValue' . $this->boundCounter;
```
Comment (greg0ire) [inline src/QueryBuilder.php:1373]: Let's switch to stof's suggestion of `:dcValue<some_param>` here please.

Comment (stof) [reply src/QueryBuilder.php:1373]: well, this is exactly why it is here (using an actual incremented number instead of my placeholder)

Comment (greg0ire) [reply src/QueryBuilder.php:1373]: Oh right I misread, sorry.

**Comment by norkunas** [comment]
PR ready again for review

**Review by greg0ire — APPROVED** [review]

**Review by SenseException — APPROVED** [review]

**Comment by greg0ire** [comment]
Thanks @norkunas !

**Pull request merged**